File: HyperFetchMobile/app/src/main/java/com/hyperfetch/rate/TokenBucketLimiter.java
package com.hyperfetch.rate;

/**
 * 令牌桶限速器实现
 * 使用令牌桶算法控制下载速率
 */
public class TokenBucketLimiter implements RateLimiter {
    
    /** 最小速率：16 KB/s = 16384 字节/秒 */
    private static final long MIN_RATE = 16 * 1024;
    
    /** 当前速率（字节/秒） */
    private volatile long rate;
    
    /** 当前桶内令牌数 */
    private volatile long tokens;
    
    /** 上次填充令牌的时间（纳秒） */
    private long lastRefillNanos;
    
    /**
     * 构造函数
     * 默认无限制速率
     */
    public TokenBucketLimiter() {
        this.rate = Long.MAX_VALUE;
        this.tokens = Long.MAX_VALUE;
        this.lastRefillNanos = System.nanoTime();
    }
    
    /**
     * 构造函数
     * 
     * @param bytesPerSecond 初始速率（字节/秒）
     */
    public TokenBucketLimiter(long bytesPerSecond) {
        setRate(bytesPerSecond);
        this.tokens = this.rate;
        this.lastRefillNanos = System.nanoTime();
    }
    
    /**
     * 申领令牌
     * 阻塞直到获取足够的令牌
     * 
     * @param bytes 需要申领的字节数
     * @throws InterruptedException 当线程被中断时抛出
     */
    @Override
    public synchronized void acquire(int bytes) throws InterruptedException {
        // 无限制速率时直接返回
        if (rate == Long.MAX_VALUE) {
            return;
        }
        
        // 循环等待直到有足够的令牌
        while (tokens < bytes) {
            refill();
            if (tokens < bytes) {
                // 计算需要等待的时间（毫秒）
                long waitMs = (bytes - tokens) * 1000 / rate;
                wait(Math.max(waitMs, 1));
            }
        }
        
        // 消耗令牌
        tokens -= bytes;
    }
    
    /**
     * 设置速率
     * 
     * @param bytesPerSecond 每秒字节数
     */
    @Override
    public synchronized void setRate(long bytesPerSecond) {
        // 如果不是无限制速率且小于最小速率，则设置为最小速率
        if (bytesPerSecond != Long.MAX_VALUE && bytesPerSecond < MIN_RATE) {
            bytesPerSecond = MIN_RATE;
        }
        
        this.rate = bytesPerSecond;
        
        // 唤醒等待的线程
        notifyAll();
    }
    
    /**
     * 获取当前速率
     * 
     * @return 当前速率（字节/秒）
     */
    @Override
    public long getRate() {
        return rate;
    }
    
    /**
     * 填充令牌
     * 根据时间差计算并添加新令牌
     */
    private void refill() {
        long now = System.nanoTime();
        // 计算应该添加的令牌数：时间差 * 速率 / 10^9
        long add = (now - lastRefillNanos) * rate / 1_000_000_000L;
        
        if (add > 0) {
            // 桶容量 = rate，令牌数不能超过桶容量
            tokens = Math.min(rate, tokens + add);
            lastRefillNanos = now;
        }
    }
}

File: temp_project/HyperFetch/app/src/main/java/com/hyperfetch/rate/TokenBucketLimiter.kt
package com.hyperfetch.rate

import java.util.concurrent.TimeUnit
import java.util.concurrent.locks.Condition
import java.util.concurrent.locks.ReentrantLock
import kotlin.math.max

/**
 * 令牌桶限速器
 * 使用令牌桶算法实现精确的速度限制
 */
class TokenBucketLimiter {

    companion object {
        private const val MIN_RATE = 16 * 1024L // 最小速率 16 KB/s
    }

    private val lock = ReentrantLock()
    private val condition: Condition = lock.newCondition()

    @Volatile
    private var rate: Long = Long.MAX_VALUE // 字节/秒，Long.MAX_VALUE 表示无限制

    private var tokens: Long = 0L
    private var lastRefillNanos: Long = System.nanoTime()

    /**
     * 获取指定字节数的令牌
     * 如果速率无限制则立即返回
     * @param bytes 需要获取的字节数
     * @throws InterruptedException 如果等待时被中断
     */
    @Throws(InterruptedException::class)
    fun acquire(bytes: Int) {
        if (rate == Long.MAX_VALUE) return

        lock.lock()
        try {
            while (tokens < bytes) {
                refill()
                if (tokens < bytes) {
                    val waitMs = max((bytes - tokens) * 1000 / rate, 1)
                    condition.await(waitMs.coerceAtMost(1000).toLong(), TimeUnit.MILLISECONDS)
                }
            }
            tokens -= bytes
        } finally {
            lock.unlock()
        }
    }

    /**
     * 尝试获取令牌，不阻塞
     * @param bytes 需要获取的字节数
     * @return true 如果获取成功，false 如果令牌不足
     */
    fun tryAcquire(bytes: Int): Boolean {
        if (rate == Long.MAX_VALUE) return true

        lock.lock()
        try {
            refill()
            if (tokens >= bytes) {
                tokens -= bytes
                return true
            }
            return false
        } finally {
            lock.unlock()
        }
    }

    /**
     * 设置限速值
     * @param bytesPerSecond 限速值（字节/秒），Long.MAX_VALUE 表示无限制
     */
    fun setRate(bytesPerSecond: Long) {
        lock.lock()
        try {
            this.rate = if (bytesPerSecond != Long.MAX_VALUE && bytesPerSecond < MIN_RATE) {
                MIN_RATE
            } else {
                bytesPerSecond
            }
            condition.signalAll()
        } finally {
            lock.unlock()
        }
    }

    /**
     * 获取当前限速值
     */
    fun getRate(): Long = rate

    /**
     * 是否启用限速
     */
    fun isLimited(): Boolean = rate != Long.MAX_VALUE

    /**
     * 补充令牌
     */
    private fun refill() {
        val now = System.nanoTime()
        val elapsed = now - lastRefillNanos
        if (elapsed > 0) {
            val add = elapsed * rate / 1_000_000_000L
            if (add > 0) {
                tokens = minOf(rate, tokens + add)
                lastRefillNanos = now
            }
        }
    }

    /**
     * 重置限速器
     */
    fun reset() {
        lock.lock()
        try {
            rate = Long.MAX_VALUE
            tokens = 0L
            lastRefillNanos = System.nanoTime()
        } finally {
            lock.unlock()
        }
    }
}
